src/control/server/harness.go

```diff
@@ -160,13 +160,21 @@ func (h *IOServerHarness) AwaitStorageReady(ctx context.Context) error {
 
 	h.log.Info("Waiting for I/O server instance storage to be ready...")
 	for _, instance := range h.instances {
-		needsSuperblock, err := instance.NeedsSuperblock()
+		needsScmFormat, err := instance.NeedsScmFormat()
 		if err != nil {
-			return err
+			return errors.Wrap(err, "failed to check storage formatting")
 		}
-		if needsSuperblock {
-			instance.AwaitStorageReady(ctx)
+
+		if !needsScmFormat {
+			needsSuperblock, err := instance.NeedsSuperblock()
+			if err != nil {
+				return errors.Wrap(err, "failed to check instance superblock")
+			}
+			if !needsSuperblock {
+				continue
+			}
 		}
+		instance.AwaitStorageReady(ctx)
```

Comment:
so if SCM is formatted we still have to AwaitStorageReady in order to format NVMe, and in that case we will get a "SCM already formatted" message when calling storage format I presume?

Reply:
The logic in this loop isn't to direct the formatting operation; it's an attempt to short-circuit the need to wait for a storage ready signal. Some comments explaining this would probably be helpful.

I added the check for SCM formatting first because checking for a superblock on unformatted storage will result in an error.

Reply:
To more directly answer your question - the code doesn't currently support formatting NVMe independently of SCM, right? So if SCM is formatted we're assuming that NVMe is also formatted. For now. Longer-term, we definitely need to support more sophisticated storage setup workflows.

Reply:
right so if  SCM is formatted then why are we awaiting storage format here?

Reply:
More precisely: If SCM is formatted AND there's a superblock present, we assume that no formatting is required. This is how the code worked before: https://github.com/daos-stack/daos/blob/87cf644ac23683fb4643d4c31d7b4f2c9a4ae4c1/src/control/server/ctl_storage_rpc.go#L159

Reply:
in the case that needsScmFormat == false but needsSuperblock == true

Reply:
> 
> 
> right so if SCM is formatted then why are we awaiting storage format here?

We're not. The logic is: if SCM is formatted, check for a superblock. If there's a superblock, continue to the next instance. The only way we get to waiting for storage ready is if either of those checks fails.

Reply:
> 
> 
> in the case that needsScmFormat == false but needsSuperblock == true

In that case, it would fall through to AwaitStorageFormat() which would receive a signal that both SCM and NVMe have been formatted and are ready to proceed, and then the harness would create the superblock later. I think that it's probably worth revisiting this soon to more completely deconflate the concepts of storage readiness and instance readiness (as determined by the presence or absence of the superblock), though. We're also going to need to support more sophisticated storage maintenance workflows (e.g. reformatting individual NVMe/SCM modules after they've been replaced due to failure, etc).